src/otx/backend/native/models/instance_segmentation/base.py

```diff
@@ -264,35 +271,104 @@ def _export_parameters(self) -> TaskLevelExportParameters:
+                self._threshold_warning_logged = True
+            return 0.5
+        return float(threshold)
+
     def on_load_checkpoint(self, ckpt: dict[str, Any]) -> None:
         """Load state_dict from checkpoint.
 
-        For detection, it is need to update confidence threshold information when
+        For instance segmentation, it is needed to update confidence threshold and F1 score information when
         the metric is FMeasure.
         """
-        if best_confidence_threshold := ckpt.get("confidence_threshold", None) or (
-            (hyper_parameters := ckpt.get("hyper_parameters", None))
-            and (best_confidence_threshold := hyper_parameters.get("best_confidence_threshold", None))
+        hyper_parameters = ckpt.get("hyper_parameters", {})
+
+        # Load best confidence threshold (legacy and new format)
+        if best_confidence_threshold := ckpt.get("confidence_threshold", None) or hyper_parameters.get(
+            "best_confidence_threshold",
+            None,
         ):
             self.hparams["best_confidence_threshold"] = best_confidence_threshold
         super().on_load_checkpoint(ckpt)
 
     def _log_metrics(self, meter: Metric, key: Literal["val", "test"], **compute_kwargs) -> None:
         if key == "val":
-            retval = super()._log_metrics(meter, key)
-
-            # NOTE: Validation metric logging can update `best_confidence_threshold`
-            if (
-                isinstance(meter, MetricCollection)
-                and (fmeasure := getattr(meter, "FMeasure", None))
-                and (best_confidence_threshold := getattr(fmeasure, "best_confidence_threshold", None))
-            ) or (
-                isinstance(meter, FMeasure)
-                and (best_confidence_threshold := getattr(meter, "best_confidence_threshold", None))
-            ):
-                self.hparams["best_confidence_threshold"] = best_confidence_threshold
-
-            return retval
+            super()._log_metrics(meter, key)
+
+            # NOTE: Only update best_confidence_threshold when we achieve a NEW best F1 score
+            fmeasure = None
+            if isinstance(meter, MetricCollection) and (fmeasure := getattr(meter, "FMeasure", None)):
+                pass  # fmeasure is set
+            elif isinstance(meter, FMeasure):
+                fmeasure = meter
+
+            if fmeasure is not None and hasattr(fmeasure, "best_confidence_threshold"):
+                # Instead of using only the last best_confidence_threshold,
+                # we now average the last N values to get a more stable estimate,
+                # especially helpful for small datasets where it can vary greatly across epochs.
+                if "best_confidence_threshold_list" not in self.hparams:
+                    self.hparams["best_confidence_threshold_list"] = []
+                self.hparams["best_confidence_threshold_list"].append(fmeasure.best_confidence_threshold)
```

Comment:
Could you explain why do we need `best_confidence_threshold_list`? Do you have some experiments to prove necessity of this average threshold? Also, could we avoid storing this list in hparams and save there only final one?

Reply:
This is explained in the comments.

Instead of relying on just the last best_confidence_threshold, we now average the last N values to obtain a more stable estimate. This is particularly useful for small datasets, where the threshold can fluctuate significantly between epochs.

The issue originates from: `https://github.com/open-edge-platform/geti/issues/940`.

Reply:
I see, how much does it increase the f1-score estimation?